File: meWrap/Controllers/WLAddContributorsViewController/WLAddressBookRecordCell/AddressBookRecordCell.m
//
//  WLrecordCell.m
//  meWrap
//
//  Created by Ravenpod on 09.05.14.
//  Copyright (c) 2014 Ravenpod. All rights reserved.
//

#import "AddressBookRecordCell.h"

@interface AddressBookRecordCell ()

@property (weak, nonatomic) IBOutlet UIButton *selectButton;
@property (nonatomic, weak) IBOutlet StreamView* streamView;
@property (nonatomic, weak) IBOutlet UILabel* nameLabel;
@property (nonatomic, weak) IBOutlet ImageView* avatarView;
@property (weak, nonatomic) IBOutlet UIButton *openView;
@property (weak, nonatomic) IBOutlet UILabel *statusLabel;
@property (weak, nonatomic) IBOutlet UILabel *pandingLabel;
@property (weak, nonatomic) IBOutlet UIButton *statusButton;
@property (strong, nonatomic) IBOutlet LayoutPrioritizer *statusPrioritizer;
@property (strong, nonatomic) StreamDataSource *dataSource;
@property (weak, nonatomic) IBOutlet UILabel *phoneLabel;


@end

@implementation AddressBookRecordCell

- (void)awakeFromNib {
    [super awakeFromNib];
    __weak typeof(self)weakSelf = self;
    self.dataSource = [[StreamDataSource alloc] initWithStreamView:self.streamView];
    [self.dataSource addMetrics:[[StreamMetrics alloc] initWithIdentifier:@"AddressBookPhoneNumberCell" initializer:^(StreamMetrics *metrics) {
        metrics.size = 50;
        metrics.selectable = YES;
        [metrics setFinalizeAppearing:^(StreamItem *item, StreamReusableView* view) {
            AddressBookPhoneNumberCell* cell = (id)view;
            AddressBookPhoneNumber *phoneNumber = item.entry;
            cell.checked = [weakSelf.delegate recordCell:weakSelf phoneNumberState:phoneNumber];
        }];
        [metrics setSelection:^(StreamItem *item, AddressBookPhoneNumber *phoneNumber) {
            [weakSelf.delegate recordCell:weakSelf didSelectPhoneNumber:phoneNumber];
        }];
    }]];
}

- (void)setup:(AddressBookRecord *)record {
	AddressBookPhoneNumber* phoneNumber = [record.phoneNumbers lastObject];
    
    User *user = phoneNumber.user;
	self.nameLabel.text = phoneNumber.name;
    NSString *url = phoneNumber.avatar.small;
    if (user && phoneNumber.activated && !url.nonempty) {
        self.avatarView.defaultBackgroundColor = Color.orange;
    } else {
        self.avatarView.defaultBackgroundColor = Color.grayLighter;
    }
    self.avatarView.url = url;
	
	if (self.streamView) {
        [self layoutIfNeeded];
        self.dataSource.items = record.phoneNumbers;
        self.statusLabel.text = @"invite_me_to_meWrap".ls;
	} else {
        self.phoneLabel.text = record.phoneStrings;
        self.pandingLabel.text = user.isInvited ? @"sign_up_pending".ls : @"";
        if (phoneNumber.activated) {
            self.statusLabel.text = @"signup_status".ls;
            
        } else if (user)  {
            self.statusLabel.text = [NSString stringWithFormat:@"invite_status".ls,
                                       [user.invitedAt stringWithDateStyle:NSDateFormatterShortStyle]];
        } else {
            self.statusLabel.text = @"invite_me_to_meWrap".ls;
        }
		self.state = [self.delegate recordCell:self phoneNumberState:phoneNumber];
        self.statusButton.hidden = !(user && (self.state == AddressBookPhoneNumberStateAdded));
        self.statusPrioritizer.defaultState = !(user && (self.state == AddressBookPhoneNumberStateAdded));
	}
}

- (void)setState:(AddressBookPhoneNumberState)state {
	_state = state;
    if (state == AddressBookPhoneNumberStateAdded) {
        self.selectButton.enabled = NO;
    } else {
        self.selectButton.enabled = YES;
        self.selectButton.selected = state == AddressBookPhoneNumberStateSelected;
    }
}

- (void)setOpened:(BOOL)opened {
	_opened = opened;
	[UIView beginAnimations:nil context:nil];
	self.openView.selected = opened;
	[UIView commitAnimations];
}

#pragma mark - Actions

- (IBAction)select:(id)sender {
    AddressBookRecord* record = self.entry;
    AddressBookPhoneNumber *person = [record.phoneNumbers lastObject];
	[self.delegate recordCell:self didSelectPhoneNumber:person];
}

- (IBAction)open:(id)sender {
	self.opened = !self.opened;
	[self.delegate recordCellDidToggle:self];
}

@end

File: meWrap/Controllers/WLAddContributorsViewController/AddressBookRecordCell/AddressBookRecordCell.swift
//
//  AddressBookRecordCell.swift
//  meWrap
//
//  Created by Yura Granchenko on 05/02/16.
//  Copyright © 2016 Ravenpod. All rights reserved.
//

import Foundation

@objc enum AddressBookPhoneNumberState: Int {
    case Default, Selected, Added
}

@objc protocol AddressBookRecordCellDelegate {
    func recordCell(cell: AddressBookRecordCell, didSelectPhoneNumber person: AddressBookPhoneNumber)
    func recordCell(cell: AddressBookRecordCell, phoneNumberState phoneNumber: AddressBookPhoneNumber) -> AddressBookPhoneNumberState
    func recordCellDidToggle(cell: AddressBookRecordCell)
}

class AddressBookRecordCell: StreamReusableView {
    
    @IBOutlet weak var delegate: AddressBookRecordCellDelegate!
    
    @IBOutlet weak var selectButton: UIButton!
    @IBOutlet weak var streamView: StreamView!
    @IBOutlet weak var nameLabel: UILabel!
    @IBOutlet weak var avatarView: ImageView!
    @IBOutlet weak var openView: UIButton!
    @IBOutlet weak var statusLabel: UILabel!
    @IBOutlet weak var pandingLabel: UILabel!
    @IBOutlet weak var statusButton: UIButton!
    @IBOutlet weak var phoneLabel: UILabel!
    @IBOutlet weak var statusPrioritizer: LayoutPrioritizer!
    var dataSource: StreamDataSource?
    
    var state: AddressBookPhoneNumberState = .Default {
        willSet {
            if newValue == .Added {
                selectButton.enabled = false
            } else {
                selectButton.enabled = true
                selectButton.selected = newValue == .Selected
            }
        }
    }
    
    var opened: Bool = false {
        willSet {
            UIView.beginAnimations(nil, context: nil)
            openView.selected = newValue
            UIView.commitAnimations()
        }
    }
    
    override func awakeFromNib() {
        super.awakeFromNib()
        if let streamView = streamView {
            dataSource = StreamDataSource(streamView: streamView)
            dataSource?.addMetrics(StreamMetrics(identifier: "AddressBookPhoneNumberCell", initializer: { metrics in
                metrics.size = 50.0
                metrics.selectable = true
                metrics.finalizeAppearing = { [weak self] item, view in
                    let cell = view as? AddressBookPhoneNumberCell
                    let phoneNumber = item.entry as? AddressBookPhoneNumber
                    if let weakSelf = self, let phoneNumber = phoneNumber {
                        cell?.checked = weakSelf.delegate.recordCell(weakSelf, phoneNumberState: phoneNumber) != .Default
                    }
                }
                metrics.selection = { [weak self] item, phoneNumber in
                    if let weakSelf = self, let phoneNumber = phoneNumber as? AddressBookPhoneNumber {
                        weakSelf.delegate.recordCell(weakSelf, didSelectPhoneNumber: phoneNumber)
                    }
                }
            }))
        }
    }
    
    override func setup(entry: AnyObject) {
        guard let record = entry as? AddressBookRecord else { return }
        guard let phoneNumber = record.phoneNumbers.last else { return }
        let url = phoneNumber.avatar?.small
        if url?.isEmpty ?? true && phoneNumber.user != nil {
            avatarView.defaultBackgroundColor = Color.orange
        } else {
            avatarView.defaultBackgroundColor = Color.grayLighter
        }
        avatarView.url = url
        

        if streamView != nil {
            layoutIfNeeded()
            dataSource?.items = record.phoneNumbers
            statusLabel.text = "invite_me_to_meWrap".ls
        } else {
            let user = phoneNumber.user
            phoneLabel.text = record.phoneStrings
            pandingLabel.text = user?.isInvited ?? false ? "sign_up_pending".ls : ""
            if phoneNumber.activated {
                statusLabel.text = "signup_status".ls
            } else if (user != nil) {
                statusLabel.text = String(format:"invite_status".ls, user?.invitedAt.stringWithDateStyle(.ShortStyle) ?? "")
            } else {
                statusLabel.text = "invite_me_to_meWrap".ls
            }
            state = delegate?.recordCell(self, phoneNumberState: phoneNumber) ?? .Default
            let notInvited = !(user != nil && state == .Added)
            statusButton?.hidden = notInvited
            statusPrioritizer?.defaultState = notInvited
        }
    }
    
    //MARK: Actions
    
     @IBAction func _select(sender: AnyObject) {
        let record = entry as? AddressBookRecord
        if let person = record?.phoneNumbers.last {
            delegate?.recordCell(self, didSelectPhoneNumber: person)
        }
    }
    
    @IBAction func open(sender: AnyObject) {
        opened = !opened
        delegate?.recordCellDidToggle(self)
    }
}